File: src/harvest/git-join.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

export async function isGitRepository(cwd: string): Promise<boolean> {
  try {
    await execFileAsync("git", ["rev-parse", "--git-dir"], { cwd, timeout: 10_000 });
    return true;
  } catch {
    return false;
  }
}

/** Nearest commit on HEAD at or before `isoTime` (exclusive of future commits). */
export async function gitCommitBefore(
  cwd: string,
  isoTime: string,
): Promise<string | null> {
  try {
    const { stdout } = await execFileAsync(
      "git",
      ["rev-list", "-1", `--before=${isoTime}`, "HEAD"],
      { cwd, timeout: 30_000 },
    );
    const sha = stdout.trim();
    return sha.length > 0 ? sha : null;
  } catch {
    return null;
  }
}

export async function gitDiffBetween(
  cwd: string,
  fromSha: string,
  toSha: string,
): Promise<string | null> {
  if (fromSha === toSha) {
    return "";
  }
  try {
    const { stdout } = await execFileAsync(
      "git",
      ["diff", `${fromSha}..${toSha}`],
      { cwd, timeout: 120_000, maxBuffer: 16 * 1024 * 1024 },
    );
    return stdout;
  } catch {
    return null;
  }
}

export async function gitShowStat(
  cwd: string,
  sha: string,
): Promise<string | null> {
  try {
    const { stdout } = await execFileAsync(
      "git",
      ["show", "--stat", "--format=", sha],
      { cwd, timeout: 60_000, maxBuffer: 4 * 1024 * 1024 },
    );
    return stdout.trim() || null;
  } catch {
    return null;
  }
}

File: python/src/curious/harvest/git_join.py
from __future__ import annotations

import subprocess
from pathlib import Path


def is_git_repository(cwd: str | Path) -> bool:
    try:
        subprocess.run(
            ["git", "rev-parse", "--git-dir"],
            cwd=cwd,
            capture_output=True,
            check=True,
            timeout=10,
        )
        return True
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, FileNotFoundError):
        return False


def git_commit_before(cwd: str | Path, iso_time: str) -> str | None:
    try:
        result = subprocess.run(
            ["git", "rev-list", "-1", f"--before={iso_time}", "HEAD"],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=True,
            timeout=30,
        )
        sha = result.stdout.strip()
        return sha if sha else None
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, FileNotFoundError):
        return None


def git_diff_between(cwd: str | Path, from_sha: str, to_sha: str) -> str | None:
    if from_sha == to_sha:
        return ""
    try:
        result = subprocess.run(
            ["git", "diff", f"{from_sha}..{to_sha}"],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=True,
            timeout=120,
        )
        return result.stdout
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, FileNotFoundError):
        return None
